mods/default/craftitems.lua

```diff
@@ -133,11 +133,14 @@ minetest.register_on_player_receive_fields(function(player, formname, fields)
 			end
 		end
 
-		local data_str = minetest.serialize(data)
-		stack:set_metadata(data_str)
+		stack:get_meta():from_table(data)
 		book_on_use(stack, player)
 	end
 
+	-- Backwards compatibility
+	stack:set_metadata("")
```

Comment:
Nope, this is required. That line is what actually fixes the main issue. `book_on_use` does not update the wielded itemstack after running `from_table`, and even if it did these changes would be overwritten back in `on_player_receive_fields` since the itemstack variable there is old and does not contain the new meta tables. This leaves that line as the best solution IMO. I can't actually update the meta in `book_on_use`, though I may be able to simply return the modified itemstack (adds 1 line compared to 2).